File: greeninference/greeninference/src/lib/physics.js
// FOPDT thermal model + PIDv2 controllers
// Ported from CoolSync_final/train_models.py

// ─── Physical constants ───────────────────────────────────────────────────────
export const PHYS = {
  RHO_AIR:    1.225,
  CP_AIR:     1006.0,
  C_RACK:     200_000.0,
  T_SUPPLY:   18.0,
  T_SETPOINT: 22.0,
  T_RL:       23.0,
  T_DANGER:   25.0,
  T_CRITICAL: 27.0,
  T_AMB:      24.0,
  DEAD_STEPS: 3,
  DT_SIM:     60.0,
  MAX_FLOW:   4.0,
  FAN_MAX_W:  5_000.0,
  FAN_MIN:    0.20,
  Q_BASE_W:   12_000.0,
  Q_RACK_MAX: 36_000.0,
  H_AMB:      30.0,
  ETA_CHILLER: 0.50,
};

const {
  RHO_AIR, CP_AIR, C_RACK, T_SUPPLY, T_SETPOINT, T_RL,
  T_DANGER, T_CRITICAL, T_AMB, DEAD_STEPS, DT_SIM,
  MAX_FLOW, FAN_MAX_W, FAN_MIN, Q_BASE_W, Q_RACK_MAX,
  H_AMB, ETA_CHILLER,
} = PHYS;

// ─── Burst class lookup (calibrated from workload_timeseries.csv) ─────────────
export const BURST_PARAMS = {
  0: { amp: 1.50, dur: 3,  label: "Short" },
  1: { amp: 1.42, dur: 7,  label: "Medium" },
  2: { amp: 1.33, dur: 13, label: "Long" },
  3: { amp: 1.25, dur: 20, label: "VeryLong" },
};

export function tokensToBurstClass(tokens) {
  if (tokens < 64)  return 0;
  if (tokens < 256) return 1;
  if (tokens < 800) return 2;
  return 3;
}

// ─── Physics helpers ──────────────────────────────────────────────────────────
export function qCool(fanPct, T, Tsup = T_SUPPLY) {
  return RHO_AIR * MAX_FLOW * fanPct * CP_AIR * Math.max(T - Tsup, 0.0);
}

export function pFan(fanPct) {
  return FAN_MAX_W * Math.pow(fanPct, 3);
}

export function pChiller(QcoolW, TsupplyC, TambC = T_AMB) {
  const TsupK = TsupplyC + 273.15;
  const TambK = TambC + 273.15;
  const COP = (TsupK / Math.max(TambK - TsupK, 0.5)) * ETA_CHILLER;
  return QcoolW / Math.max(COP, 1.0);
}

export function rackStep(T, Qdelayed, fanPct, Tsup = T_SUPPLY, dt = DT_SIM) {
  const Qc  = qCool(fanPct, T, Tsup);
  const Ql  = H_AMB * (T - T_AMB);
  const dT  = ((Qdelayed - Qc - Ql) / C_RACK) * dt;
  const Tnew = Math.max(10.0, Math.min(55.0, T + dT));
  return { Tnew, Qc, Pfan: pFan(fanPct) };
}

// ─── Controllers ──────────────────────────────────────────────────────────────
export class PIDv2 {
  constructor({
    kp = 0.015, ki = 0.001, sp = T_SETPOINT,
    fanMin = FAN_MIN, fanMax = 1.0,
    windup = 400.0, deadband = 1.5,
  } = {}) {
    this.kp = kp; this.ki = ki;
    this.sp = sp; this.fanMin = fanMin; this.fanMax = fanMax;
    this.windup = windup; this.deadband = deadband;
    this._I = 0.0; this._fanOut = 0.61;
  }

  reset() { this._I = 0.0; this._fanOut = 0.61; }

  compute(T, dt = DT_SIM) {
    const e = T - this.sp;
    if (Math.abs(e) > this.deadband) {
      this._I = Math.max(-this.windup, Math.min(this.windup, this._I + e * dt));
    } else {
      this._I *= 0.7;
    }
    const fanRaw     = 0.61 + this.kp * e + this.ki * this._I;
    const fanClipped = Math.max(this.fanMin, Math.min(this.fanMax, fanRaw));
    const alpha      = T < T_DANGER ? 0.30 : 0.80;
    this._fanOut     = alpha * fanClipped + (1.0 - alpha) * this._fanOut;
    return this._fanOut;
  }
}

export class PIDv2Conservative {
  static T_SUPPLY_COLD = 16.0;

  constructor({
    kp = 0.015, ki = 0.001, sp = 20.0,
    fanMin = FAN_MIN, fanMax = 1.0,
    windup = 400.0, deadband = 0.5,
  } = {}) {
    this.kp = kp; this.ki = ki;
    this.sp = sp; this.fanMin = fanMin; this.fanMax = fanMax;
    this.windup = windup; this.deadband = deadband;
    this._I = 0.0;
  }

  reset() { this._I = 0.0; }

  compute(T, dt = DT_SIM) {
    const e = T - this.sp;
    if (Math.abs(e) > this.deadband) {
      this._I = Math.max(-this.windup, Math.min(this.windup, this._I + e * dt));
    } else {
      this._I *= 0.9;
    }
    const fanRaw = 0.61 + this.kp * e + this.ki * this._I;
    return Math.max(this.fanMin, Math.min(this.fanMax, fanRaw));
  }
}

// ─── Workload generator ───────────────────────────────────────────────────────
export function generateWorkload(nSteps, burstStart, burstClass, concurrentUsers = 1) {
  const heat = new Array(nSteps).fill(Q_BASE_W);
  const bp   = BURST_PARAMS[burstClass] ?? BURST_PARAMS[1];
  const Qrack = Math.min(Q_BASE_W * (1 + (bp.amp - 1) * concurrentUsers), Q_RACK_MAX);
  const end  = Math.min(burstStart + bp.dur, nSteps);
  for (let i = burstStart; i < end; i++) heat[i] = Qrack;
  return heat;
}

// ─── Core simulation loop ─────────────────────────────────────────────────────
function runLoop(heatTrace, getAction, T0, Tsup0) {
  const n      = heatTrace.length;
  const deadBuf = new Array(DEAD_STEPS).fill(heatTrace[0] * 0.5);
  let T    = T0;
  let cumE = 0.0;
  const rows = [];

  for (let step = 0; step < n; step++) {
    const Q = heatTrace[step];
    const { fan, tsup } = getAction(T, step);
    const { Tnew, Qc, Pfan } = rackStep(T, deadBuf[0], fan, tsup);
    deadBuf.shift(); deadBuf.push(Q);
    const Pchill = pChiller(Qc, tsup);
    const Ptotal = Pfan + Pchill;
    cumE += Ptotal * DT_SIM / 3600.0;

    // phase based on heat level relative to baseline
    let phase = "RECOVERY";
    if (Q > Q_BASE_W * 1.05) phase = "DECODE";
    else if (step < DEAD_STEPS) phase = "PREFILL";

    rows.push({
      t:           step * 60_000,
      step,
      inlet:       T,
      T_rack:      T,
      fan_pct:     fan,
      fanPct:      fan,
      P_fan_w:     Pfan,
      P_chiller_w: Pchill,
      P_total_w:   Ptotal,
      Q_gpu:       Q,
      heatW:       Q,
      gpuW:        Q,
      energy_wh:   cumE,
      T_supply:    tsup,
      tSupply:     tsup,
      coolingKw:   (Pfan + Pchill) / 1000,
      phase,
    });
    T = Tnew;
  }
  return rows;
}

// ─── Strategy runners ─────────────────────────────────────────────────────────
export function runPID(heatTrace) {
  const pid = new PIDv2();
  return runLoop(
    heatTrace,
    (T) => ({ fan: pid.compute(T), tsup: T_SUPPLY }),
    T_SETPOINT,
    T_SUPPLY,
  );
}

export function runPIDConservative(heatTrace) {
  const pid  = new PIDv2Conservative();
  const tsup = PIDv2Conservative.T_SUPPLY_COLD;
  return runLoop(
    heatTrace,
    (T) => ({ fan: pid.compute(T), tsup }),
    20.0,
    tsup,
  );
}

export function runCoordinated(heatTrace) {
  // Simulates LSTM+DQN lookahead benefit: fan ramps DEAD_STEPS before burst arrives.
  // Pre-cooling fires when predicted future heat exceeds threshold (= LSTM signal).
  const pid = new PIDv2();
  const n   = heatTrace.length;
  return runLoop(
    heatTrace,
    (T, step) => {
      const futureQ = heatTrace[Math.min(step + DEAD_STEPS, n - 1)];
      const pid_fan = pid.compute(T);
      if (futureQ > Q_BASE_W * 1.15 && T < T_DANGER) {
        return {
          fan:  Math.min(1.0, pid_fan + 0.15),
          tsup: 19.0,
        };
      }
      return { fan: pid_fan, tsup: T_SUPPLY };
    },
    T_SETPOINT,
    T_SUPPLY,
  );
}

// ─── High-level helpers for simulate.js ──────────────────────────────────────
export function buildPhysicsRun({ tokens = 128, strategy = "reactive" }) {
  const burstClass  = tokensToBurstClass(tokens);
  const bp          = BURST_PARAMS[burstClass];
  const nSteps      = Math.max(bp.dur + 20, 30); // enough pre- and post-burst steps
  const burstStart  = 8;

  const heatTrace = generateWorkload(nSteps, burstStart, burstClass, 1);

  let points;
  if (strategy === "coordinated") {
    points = runCoordinated(heatTrace);
  } else {
    points = runPID(heatTrace);
  }

  const energyWh    = points[points.length - 1]?.energy_wh ?? 0;
  const peakT       = Math.max(...points.map(p => p.T_rack));
  const breachSteps = points.filter(p => p.T_rack > T_DANGER).length;
  // Coordinated mode: LSTM prediction eliminates effective cooling lag
  const lagMs = strategy === "coordinated" ? 0 : DEAD_STEPS * 60_000;

  return { points, energyWh, peakT, breachSteps, lagMs };
}

export function buildConservativeRun({ tokens = 128 }) {
  const burstClass  = tokensToBurstClass(tokens);
  const bp          = BURST_PARAMS[burstClass];
  const nSteps      = Math.max(bp.dur + 20, 30);
  const burstStart  = 8;

  const heatTrace = generateWorkload(nSteps, burstStart, burstClass, 1);
  const points    = runPIDConservative(heatTrace);

  const energyWh    = points[points.length - 1]?.energy_wh ?? 0;
  const peakT       = Math.max(...points.map(p => p.T_rack));
  const breachSteps = points.filter(p => p.T_rack > T_DANGER).length;

  return { points, energyWh, peakT, breachSteps, lagMs: DEAD_STEPS * 60_000 };
}

File: backend/physics.py
"""
CoolSync physics engine — ported from CoolSync_final/train_models.py.
Single source of truth for server-side simulation (mirrors src/lib/physics.js).
"""
from __future__ import annotations
import numpy as np
from collections import deque

# ─── Constants ───────────────────────────────────────────────────────────────
RHO_AIR    = 1.225
CP_AIR     = 1006.0
C_RACK     = 200_000.0
T_SUPPLY   = 18.0
T_SETPOINT = 22.0
T_RL       = 23.0
T_DANGER   = 25.0
T_CRITICAL = 27.0
T_AMB      = 24.0
DEAD_STEPS = 3
DT_SIM     = 60.0
MAX_FLOW   = 4.0
FAN_MAX_W  = 5_000.0
FAN_MIN    = 0.20
Q_BASE_W   = 12_000.0
Q_RACK_MAX = 36_000.0
H_AMB      = 30.0
ETA_CHILLER = 0.50

FAN_LEVELS  = np.array([0.30, 0.45, 0.60, 0.75, 1.00])
TSUP_LEVELS = np.array([16.0, 19.0])
JOINT_ACTIONS = [(float(f), float(ts)) for f in FAN_LEVELS for ts in TSUP_LEVELS]
N_ACTIONS   = len(JOINT_ACTIONS)   # 10

BURST_PARAMS = {
    0: dict(amp=1.50, dur=3,  label="Short"),
    1: dict(amp=1.42, dur=7,  label="Medium"),
    2: dict(amp=1.33, dur=13, label="Long"),
    3: dict(amp=1.25, dur=20, label="VeryLong"),
}

# ─── Physics helpers ──────────────────────────────────────────────────────────
def q_cool(fan_pct: float, T: float, T_sup: float = T_SUPPLY) -> float:
    return RHO_AIR * MAX_FLOW * fan_pct * CP_AIR * max(T - T_sup, 0.0)

def p_fan(fan_pct: float) -> float:
    return FAN_MAX_W * fan_pct ** 3

def p_chiller(Q_cool_W: float, T_supply_C: float, T_amb_C: float = T_AMB) -> float:
    T_sup_K = T_supply_C + 273.15
    T_amb_K = T_amb_C    + 273.15
    COP = T_sup_K / max(T_amb_K - T_sup_K, 0.5) * ETA_CHILLER
    return Q_cool_W / max(COP, 1.0)

def rack_step(T: float, Q_delayed: float, fan_pct: float,
              T_sup: float = T_SUPPLY, dt: float = DT_SIM):
    Qc  = q_cool(fan_pct, T, T_sup)
    Ql  = H_AMB * (T - T_AMB)
    dT  = (Q_delayed - Qc - Ql) / C_RACK * dt
    return float(np.clip(T + dT, 10.0, 55.0)), Qc, p_fan(fan_pct)

# ─── Controllers ──────────────────────────────────────────────────────────────
class PIDv2:
    def __init__(self, kp=0.015, ki=0.001, sp=T_SETPOINT,
                 fan_min=FAN_MIN, fan_max=1.0, windup=400.0, deadband=1.5):
        self.kp, self.ki = kp, ki
        self.sp = sp; self.fan_min = fan_min; self.fan_max = fan_max
        self.windup = windup; self.deadband = deadband
        self._I = 0.0; self._fan_out = 0.61

    def reset(self): self._I = 0.0; self._fan_out = 0.61

    def compute(self, T, dt=DT_SIM):
        e = T - self.sp
        if abs(e) > self.deadband:
            self._I = float(np.clip(self._I + e * dt, -self.windup, self.windup))
        else:
            self._I *= 0.7
        fan_raw     = 0.61 + self.kp * e + self.ki * self._I
        fan_clipped = float(np.clip(fan_raw, self.fan_min, self.fan_max))
        alpha       = 0.30 if T < T_DANGER else 0.80
        self._fan_out = alpha * fan_clipped + (1.0 - alpha) * self._fan_out
        return self._fan_out


class PIDv2Conservative:
    T_SUPPLY_COLD = 16.0

    def __init__(self, kp=0.015, ki=0.001, sp=20.0,
                 fan_min=FAN_MIN, fan_max=1.0, windup=400.0, deadband=0.5):
        self.kp, self.ki = kp, ki
        self.sp = sp; self.fan_min = fan_min; self.fan_max = fan_max
        self.windup = windup; self.deadband = deadband
        self._I = 0.0

    def reset(self): self._I = 0.0

    def compute(self, T, dt=DT_SIM):
        e = T - self.sp
        if abs(e) > self.deadband:
            self._I = float(np.clip(self._I + e * dt, -self.windup, self.windup))
        else:
            self._I *= 0.9
        fan_raw = 0.61 + self.kp * e + self.ki * self._I
        return float(np.clip(fan_raw, self.fan_min, self.fan_max))

# ─── Workload generator ───────────────────────────────────────────────────────
def generate_workload(n_steps: int, burst_start: int = 8,
                      burst_class: int = 1, concurrent_users: int = 1):
    heat = np.full(n_steps, Q_BASE_W, dtype=float)
    bp   = BURST_PARAMS[burst_class]
    Q_rack = min(Q_BASE_W * (1 + (bp["amp"] - 1) * concurrent_users), Q_RACK_MAX)
    end  = min(burst_start + bp["dur"], n_steps)
    heat[burst_start:end] = Q_rack
    return heat

# ─── Simulation runners ───────────────────────────────────────────────────────
def _run_loop(heat_trace, get_action, T0=T_SETPOINT, tsup0=T_SUPPLY):
    buf   = deque([heat_trace[0] * 0.5] * DEAD_STEPS, maxlen=DEAD_STEPS)
    T     = T0
    cum_e = 0.0
    rows  = []
    n     = len(heat_trace)

    for step, Q in enumerate(heat_trace):
        fan, tsup = get_action(T, step)
        T_new, Qc, Pf = rack_step(T, buf[0], fan, tsup)
        buf.append(Q)
        Pc     = p_chiller(Qc, tsup)
        Pt     = Pf + Pc
        cum_e += Pt * DT_SIM / 3600.0

        if   Q > Q_BASE_W * 1.05: phase = "DECODE"
        elif step < DEAD_STEPS:    phase = "PREFILL"
        else:                      phase = "RECOVERY"

        rows.append({
            "t":           step * 60_000,
            "step":        step,
            "T_rack":      round(T, 4),
            "inlet":       round(T, 4),
            "fan_pct":     round(fan, 4),
            "P_fan_w":     round(Pf, 2),
            "P_chiller_w": round(Pc, 2),
            "P_total_w":   round(Pt, 2),
            "Q_gpu":       round(Q, 1),
            "gpuW":        round(Q, 1),
            "heatW":       round(Q, 1),
            "energy_wh":   round(cum_e, 4),
            "T_supply":    round(tsup, 2),
            "tSupply":     round(tsup, 2),
            "coolingKw":   round((Pf + Pc) / 1000, 4),
            "phase":       phase,
        })
        T = T_new

    return rows


def run_pid(heat_trace):
    pid = PIDv2()
    return _run_loop(heat_trace, lambda T, _: (pid.compute(T), T_SUPPLY),
                     T0=T_SETPOINT, tsup0=T_SUPPLY)


def run_pid_conservative(heat_trace):
    pid  = PIDv2Conservative()
    tsup = PIDv2Conservative.T_SUPPLY_COLD
    return _run_loop(heat_trace, lambda T, _: (pid.compute(T), tsup),
                     T0=20.0, tsup0=tsup)


def run_coordinated(heat_trace):
    """Simulates LSTM lookahead: fan ramps DEAD_STEPS before burst arrives."""
    pid = PIDv2()
    n   = len(heat_trace)

    def get_action(T, step):
        future_Q = heat_trace[min(step + DEAD_STEPS, n - 1)]
        pid_fan  = pid.compute(T)
        if future_Q > Q_BASE_W * 1.15 and T < T_DANGER:
            return min(1.0, pid_fan + 0.15), 19.0
        return pid_fan, T_SUPPLY

    return _run_loop(heat_trace, get_action, T0=T_SETPOINT, tsup0=T_SUPPLY)


def run_dqn(heat_trace, dqn_model, lstm_model, joint_actions, device):
    """Full DQN+LSTM inference using loaded PyTorch models."""
    import torch
    ACT_HOLD   = 2
    LOOKBACK   = DEAD_STEPS
    buf        = deque([heat_trace[0] * 0.5] * DEAD_STEPS, maxlen=DEAD_STEPS)
    T, fan, T_sup = T_RL, 0.65, 19.0
    cum_e     = 0.0
    T_hist    = [T]
    Q_hist    = [float(heat_trace[0])]
    fan_hist  = [fan]
    hold_counter = 0
    rows = []
    n    = len(heat_trace)

    def lstm_predict(T_h, Q_h, f_h):
        if len(T_h) < LOOKBACK:
            return (T_h[-1] - 22.0) / 5.0
        seq_T   = np.array([(t - 22.0) / 5.0  for t in T_h[-LOOKBACK:]])
        seq_Q   = np.array([(q - Q_BASE_W) / Q_BASE_W for q in Q_h[-LOOKBACK:]])
        seq_fan = np.array(f_h[-LOOKBACK:])
        x = torch.tensor(
            np.stack([seq_T, seq_Q, seq_fan], axis=1),
            dtype=torch.float32,
        ).unsqueeze(0).to(device)
        with torch.no_grad():
            return float(lstm_model(x).item())

    def make_obs(T_cur, Q_now, fan_p, q_buf_cur, step, total, T_pred_n, tsup_p):
        buf_list    = list(q_buf_cur)
        t_err       = (T_cur - T_RL) / 5.0
        q_now       = (Q_now - Q_BASE_W) / Q_BASE_W
        f_prev      = (fan_p - 0.5) / 0.5
        q_mean      = (float(np.mean(buf_list)) - Q_BASE_W) / Q_BASE_W
        q_trnd      = (Q_now - buf_list[-1]) / Q_BASE_W
        above       = max(0.0, T_cur - T_RL) / 5.0
        below       = max(0.0, T_RL - T_cur) / 5.0
        phase       = step / max(total - 1, 1)
        q_lag       = (buf_list[0] - Q_BASE_W) / Q_BASE_W
        T_pred_d    = (T_pred_n * 5.0) - (T_cur - 22.0)
        tsup_norm   = (tsup_p - 19.0) / 3.0
        T_sup_K     = tsup_p + 273.15
        cop         = T_sup_K / max((T_AMB + 273.15) - T_sup_K, 0.5) * ETA_CHILLER
        cop_norm    = (cop - 3.0) / 2.0
        return np.array([
            t_err, q_now, f_prev, q_mean, q_trnd,
            above, below, phase, q_lag,
            T_pred_n, T_pred_d / 5.0,
            tsup_norm, cop_norm,
        ], dtype=np.float32)

    import torch
    for step, Q in enumerate(heat_trace):
        T_pred = lstm_predict(T_hist, Q_hist, fan_hist)
        obs    = make_obs(T, Q, fan, buf, step, n, T_pred, T_sup)

        if hold_counter <= 0 or T >= T_DANGER:
            obs_t = torch.tensor(obs, dtype=torch.float32).unsqueeze(0).to(device)
            with torch.no_grad():
                a = int(dqn_model(obs_t).argmax())
            new_fan, new_tsup = joint_actions[a]
            if new_fan != fan or new_tsup != T_sup:
                fan, T_sup = new_fan, new_tsup
                hold_counter = ACT_HOLD
            else:
                hold_counter = max(hold_counter - 1, 0)
        else:
            hold_counter -= 1

        T_new, Qc, Pf = rack_step(T, buf[0], fan, T_sup)
        buf.append(Q)
        Pc     = p_chiller(Qc, T_sup)
        Pt     = Pf + Pc
        cum_e += Pt * DT_SIM / 3600.0

        phase = "DECODE" if Q > Q_BASE_W * 1.05 else ("PREFILL" if step < DEAD_STEPS else "RECOVERY")

        rows.append({
            "t":           step * 60_000,
            "step":        step,
            "T_rack":      round(T, 4),
            "inlet":       round(T, 4),
            "fan_pct":     round(fan, 4),
            "P_fan_w":     round(Pf, 2),
            "P_chiller_w": round(Pc, 2),
            "P_total_w":   round(Pt, 2),
            "Q_gpu":       round(Q, 1),
            "gpuW":        round(Q, 1),
            "heatW":       round(Q, 1),
            "energy_wh":   round(cum_e, 4),
            "T_supply":    round(T_sup, 2),
            "tSupply":     round(T_sup, 2),
            "coolingKw":   round((Pf + Pc) / 1000, 4),
            "T_pred":      round(T_pred * 5.0 + 22.0, 3),
            "phase":       phase,
        })
        T = T_new
        T_hist.append(T_new)
        Q_hist.append(Q)
        fan_hist.append(fan)

    return rows


def summarise(rows):
    energy_wh    = rows[-1]["energy_wh"] if rows else 0.0
    peak_T       = max(r["T_rack"] for r in rows) if rows else 0.0
    breach_steps = sum(1 for r in rows if r["T_rack"] > T_DANGER)
    fan_diffs    = [abs(rows[i]["fan_pct"] - rows[i-1]["fan_pct"])
                    for i in range(1, len(rows))]
    fan_var      = float(np.mean(fan_diffs)) if fan_diffs else 0.0
    return {
        "energy_wh":    round(energy_wh, 3),
        "peak_T":       round(peak_T, 3),
        "breach_steps": breach_steps,
        "fan_var":      round(fan_var, 5),
    }
